vk: Resource binding model rewrites [part 2 of 2]

kd-11 opened the pull request:
Relates to https://github.com/RPCS3/rpcs3/issues/17277

### Changes
- Remove all per-draw UBO objects and replace with SSBO array-of-structs model. This removes a lot of descriptor bind operations completely.
- Avoid using more than a few bytes of data in push constants. Instead, we write a per-draw config and push a pointer offset to the next one. Restricting push constants memory to 4 bytes per draw has decent gains.
- Replace std::vector with rsx::simple_array for trivial storage. The latter has small-vector optimizations that drastically reduce the number of memory allocations done on the fly and gives a decent perf bump to both Vulkan and OpenGL.
- Other minor optimizations related to resource bindings.

Overall, a 10-20% boost is observed depending on the game. MESA drivers benefit the most here since they doesn't keep a state cache like the properietary ones so state changes are more expensive on those drivers.

digant73 commented:
quick test on KZ3 but perf seems worse (more gpu usage and lower fps).

### Current build:

![kz3_current](https://github.com/user-attachments/assets/2877adbf-9456-4f5e-a3f7-fa8cd18d4463)

### PR:

![kz3_pr](https://github.com/user-attachments/assets/28abaf76-0aa6-4972-9581-8d948869306d)

AniLeo commented:
No system specifications, no cpu/gpu/driver info, no performance overlay with averages, can't do anything just from that

Darkhost1999 commented:
Battlefield 3
```
F {SPU[0x0000600] Thread (JM_SPU5) [0x06490]} SIG: Thread terminated due to fatal error: Assertion Failed! Vulkan API call failed with unrecoverable error: Device lost (Driver crashed with unspecified error or stopped responding and recovered) (VK_ERROR_DEVICE_LOST)
---------------- EXTRA INFORMATION --------------------
Device Fault Information:
Fault Summary:
  Address Fault Information:
    Fault at address 0xd06921000 caused by access_write
(in file C:\a\rpcs3\rpcs3\rpcs3\Emu\RSX\VK\vkutils\sync.cpp:610[:5], in function 'VkResult vk::wait_for_event(vk::event *,unsigned __int64)')
(in file C:\a\rpcs3\rpcs3\rpcs3\Emu\RSX\VK\vkutils\shared.cpp:205[:24], in function 'void vk::die_with_error()')
```
[PR.log](https://github.com/user-attachments/files/21708446/RPCS3.log)
Master works and log is as follows
[Master.log](https://github.com/user-attachments/files/21708448/RPCS3.log)

Megamouse commented on rpcs3/Emu/RSX/Program/FragmentProgramDecompiler.cpp:
```diff
@@ -1,5 +1,8 @@
 #include "stdafx.h"
+
+#pragma optimize("", off)
```
Leftover?

gamrXerus commented:
I tested 10 games. (Call of Duty Black Ops, CN Punchtime Explosion, Goldeneye Reloaded, Playstation All Stars Battle Royale, Street Fighter X Tekken, Super Street Fighter 2 Turbo, Need for Speed Carbon, Tekken 6 Virtua Fighter 5 and WWE All Stars). They all went ingame without crashes or graphical glitches. 4 player splitscreen with Call of Duty i saw the fps go from approximately 15fps average to 18 fps average. So all good on my end so far.
Master:                                                                                   
<img width="1024" height="556" alt="master" src="https://github.com/user-attachments/assets/81d597f9-c965-4768-b754-e92876d2839f" />
PR:
<img width="1024" height="563" alt="pr" src="https://github.com/user-attachments/assets/e5e361a4-c662-4fda-901c-7040cde9c095" />

digant73 commented:
Got a DEVICE_LOST in Resistance 3 after few minutes of gameplay.
Attached also the log file.

**NOTE:** I confirm also the worse perf on KZ3 (previous post edited adding onscreen stats

[RPCS3.zip](https://github.com/user-attachments/files/21721609/RPCS3.zip)

<img width="3537" height="504" alt="image" src="https://github.com/user-attachments/assets/3fcd821a-9da6-476f-b541-752118df723d" />

Darkhost1999 commented:
> Got a DEVICE_LOST in Resistance 3 after few minutes of gameplay.

Is this unlike my issue and able to replicated/consistent?
Can someone encounter the same error on a specific level triggered the same way with debug output enabled?

kd-11 commented:
I think the perf loss is just because I forgot to restore optimizations in the shader compiler.

digant73 commented:
Tested also GT6 is a lot better on this PR. Avg of 41 fps on current release against solid 60 fps on PR (even more than 75 fps if I disable vsync on GPU tab). But the PR seems to make a sort of tearing in the center part of the screen (especially when cars are in front of you

### Current build:

![gt6_current](https://github.com/user-attachments/assets/13456017-9b33-4d83-a19a-d1e43ae81ac0)

### PR:

![gt6_pr](https://github.com/user-attachments/assets/0b8aaa9e-4b5e-4dc6-8281-d63a25aa04b9)

digant73 commented:
Infamous 2 demo crashing in the same place with the following error just when shooting the first enemy in the same place:

[RPCS3.zip](https://github.com/user-attachments/files/21722736/RPCS3.zip)

<img width="1916" height="373" alt="image" src="https://github.com/user-attachments/assets/6ef0c551-65d5-4752-85eb-bcd917265698" />

![infamous_2_demo](https://github.com/user-attachments/assets/39cc9302-34b2-4609-93e0-f85682a2accb)

digant73 commented:
same crash as in Infamous 2 demo also in UC1 just in a place where current build has an avg of 40 fps (on top of the u-boat:

<img width="2243" height="334" alt="image" src="https://github.com/user-attachments/assets/764ca188-acc9-403d-a23f-1b830dd4e777" />

![uc1_pr_crash](https://github.com/user-attachments/assets/e58ff95e-919e-41a3-9d81-a2ca90ae01dd)

kd-11 commented:
The crashes are fixed so I pushed the commits though I have drafted the PR due to serious issues with vulkan shader interpreter.

digant73 commented:
> I think the perf loss is just because I forgot to restore optimizations in the shader compiler.

Hmm no difference on perf on KZ3 with latest commit

kd-11 commented:
> > I think the perf loss is just because I forgot to restore optimizations in the shader compiler.
> 
> Hmm no difference on perf on KZ3 with latest commit. Also the crashes on shaders are still there

Compilation crashes in the log you provided previously should be fixed already. If there are new crashes, upload the log files.

Darkhost1999 commented:
[RPCS3.log](https://github.com/user-attachments/files/21735530/RPCS3.log)
Here's the log for farcry 3

Megamouse commented on rpcs3/Emu/RSX/Program/FragmentProgramDecompiler.cpp:
```diff
@@ -232,23 +234,26 @@ std::string FragmentProgramDecompiler::AddCond()
+	else
 	{
-		return name;
+		index =::size32(properties.constant_offsets);
```
```suggestion
		index = ::size32(properties.constant_offsets);
```

kd-11 commented:
> > > > I think the perf loss is just because I forgot to restore optimizations in the shader compiler.
> > > 
> > > 
> > > Hmm no difference on perf on KZ3 with latest commit. Also the crashes on shaders are still there
> > 
> > 
> > Compilation crashes in the log you provided previously should be fixed already. If there are new crashes, upload the log files.
> 
> Here the log for Infamous 2 demo. It crashes in the same place
> 
> [RPCS3.log](https://github.com/user-attachments/files/21733581/RPCS3.log)
> 
> <img alt="image" width="1972" height="522" src="https://private-user-images.githubusercontent.com/64427768/476986100-8ae0728c-a73c-4d41-b330-dd236fcb9d2c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUwMzg0MzIsIm5iZiI6MTc1NTAzODEzMiwicGF0aCI6Ii82NDQyNzc2OC80NzY5ODYxMDAtOGFlMDcyOGMtYTczYy00ZDQxLWIzMzAtZGQyMzZmY2I5ZDJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODEyVDIyMzUzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwOTM4NTg1YWQ3MDdjNjM2NDFlODQ0ODVmOTk2MTU5N2E1YTExNmI5N2JiZDI0ZTA1ZTcyZjczMDFiOTM5NDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zci1VtAoyHf9jJD99YDC6RmfoiujxposEZWowQl8sjg">

Looks like a glslang bug? Same shader compiles fine for me. I'll circle back to this once I finish the interpreter work.

digant73 commented:
with latest two commits the crashes are still present even in other places were it was not previously present such as in the following, in UC1 (u-boat chapter):

[RPCS3.zip](https://github.com/user-attachments/files/21751563/RPCS3.zip)

<img width="3839" height="2159" alt="image" src="https://github.com/user-attachments/assets/a9e09cc6-532f-45dd-8989-41ec717f92de" />

digant73 commented:
Batman arkham city demo. Glicthes present in the PR:

<img width="3838" height="2159" alt="image" src="https://github.com/user-attachments/assets/862ed975-604b-4f21-8c9b-306e39875b9e" />

EmulationChannell commented:
<img width="1911" height="1028" alt="Captura de tela 2025-08-14 185449" src="https://github.com/user-attachments/assets/2e9357ba-579b-4a34-859c-1e2647b79c1a" />
[RPCS3.log](https://github.com/user-attachments/files/21780949/RPCS3.log)

kd-11 commented:
Compiler crash is definitely a glslang bug. I'll figure out a workaround.

bigol83 commented:
GTA IV also fixed for me, performance is about the same, maybe slightly lower in PR compared to master, like 1-2 fps lower in PR compared to master. I am on rtx 3060 gpu

kd-11 commented:
Not sure what is bugging with the NV driver, definitely something from their end. CPU usage should be much lower, and I don't have any plans on moving stuff back to the CPU. In fact, the roadmap offloads even more work to the GPU. I'll benchmark a few things, I probably just need to remove the UAVs from the FS stage. I'll see what nsight has to say about it.

digant73 commented:
> Not sure what is bugging with the NV driver, definitely something from their end. CPU usage should be much lower, and I don't have any plans on moving stuff back to the CPU. In fact, the roadmap offloads even more work to the GPU. I'll benchmark a few things, I probably just need to remove the UAVs from the FS stage. I'll see what nsight has to say about it.

ok. CPU usage remained basically the same in my 9950X3D (otherwise I also reported it). Only GPU usage is worse

kd-11 commented:
NV driver doesn't like back-to-back dependent SSBO reads. I'll come up with a workaround to fix the perf drop.

Valtekken commented:
> > > Not sure what is bugging with the NV driver, definitely something from their end. CPU usage should be much lower, and I don't have any plans on moving stuff back to the CPU. In fact, the roadmap offloads even more work to the GPU. I'll benchmark a few things, I probably just need to remove the UAVs from the FS stage. I'll see what nsight has to say about it.
> > 
> > 
> > Offloading work to the GPU sounds good, would this result in lower requirements on the CPU side by any chance?
> 
> Yes.

Great, thanks for your hard work!

kd-11 commented:
@digant73 Try the latest version.

Asinin3 commented:
KZ3 100% res scale and with MLAA Disabled via patch

Master
<img width="2564" height="1500" alt="image" src="https://github.com/user-attachments/assets/a3b56778-73cb-466d-bae6-aaa24a4c5b75" />
Master in best light (captured after waiting a really long time and for graph to have no fluctuations)
<img width="2564" height="1500" alt="image" src="https://github.com/user-attachments/assets/176ea560-9767-49a9-8f8f-2244abdd5cdd" />

PR (waited awhile to show in best light again)
<img width="2564" height="1500" alt="image" src="https://github.com/user-attachments/assets/7fc7557c-85e0-41a9-a27d-6a7c25ff8c21" />

kd-11 commented:
It's a GPU issue on vertex shader. The only reason it manifests for some users is that their GPUs were pretty close to getting maxed out. In my case, my GPU is very low end.
I was able to get nsight to isolate the rendering code in KZ3 so that I could do GPU-only benchmark using the tools without rpcs3 in the way and the vertex shaders are way heavier on ampere. The GPU is doing something wrong for sure, so I have to work around that for now. My AMD card doesn't seem to care though.

woj1993 commented:
> It's a GPU issue on vertex shader. The only reason it manifests for some users is that their GPUs were pretty close to getting maxed out. In my case, my GPU is very low end. I was able to get nsight to isolate the rendering code in KZ3 so that I could do GPU-only benchmark using the tools without rpcs3 in the way and the vertex shaders are way heavier on ampere. The GPU is doing something wrong for sure, so I have to work around that for now. My AMD card doesn't seem to care though.

Witch issue as we have rendering issues in this branch and performance?

kd-11 commented:
> Witch issue as we have rendering issues in this branch and performance?

There is no rendering issue with newer build artifacts. Please ensure you are on the latest artifacts version, some previous commits in the list have some graphical anomalies.

kd-11 commented:
KZ3 slowdown is caused by a feedback loop. CPU is actually faster but GPU is slower. However, the gain in CPU is lost because the game uses occlusion queries (turning off in debug settings does not help) and the round-trip-time is significantly higher. This causes the game to stall  longer than before lowering performance by quite a bit.
Unfortunately nsight isn't very useful since nvidia will never disclose their actual ISA, so it's now a guessing game which changes can improve performance. Worst case, I'll just blacklist the cards from using the optimal path, though not even sure how to maintain that without splitting the code into 2 vulkan backends.

kd-11 commented:
> I still have graphical corruption in Asura's Wrath that is not on main branch

Log?

willowdale16 commented:
For some reason, the latest commit here lost the perf improvement i got compared to commit from 4 days ago.
<img width="1079" height="227" alt="Screenshot_20250821-224150618 (1)" src="https://github.com/user-attachments/assets/e145d220-4da6-4768-b929-2bfb422c8879" />

That commit above is massive. From 52fps to 60fps on GOW Ascension.

kd-11 commented:
@woj1993 Clang is unofficial. Fixing any issues coming from that right now is a secondary concern. For all we know the bug could be coming from a compiler problem.

woj1993 commented:
@kd-11 So you don't have any compiler that is stable as Visual Studio has bug in SIMD Code optimization. That is also a reason that RPCS3 team has locked RPCS3 version to old build by default. I want to mention that this bug also happens on https://github.com/RPCS3/rpcs3/actions/runs/17055460027/job/48352101417?pr=17427 but I see you can't reproduce the error and are ignoring CLANG builds while it is known that Visual Studio builds have compiler errors https://github.com/RPCS3/rpcs3/issues/17326 so I don't know how I can help anymore. I definitely not going to retest Visual Studio Builds as they are not deterministic. I have checked that on Demon Souls where I was randomly having audio related bugs not existing on CLANG builds that could happen but also couldn't like voices missing but only sometimes or broken sound but only sometimes so I'm out of option if I won't just compile it myself using some compiler you will mark as supported that also isn't known to be broken but I don't want to read licenses to compiler, dependencies, install it all in my computer to build RPCS3 one time to test 1 game that might have an error that you won't reproduce

digant73 commented:
@woj1993 We all are using MSVC to test PRs because it is still more safe than clang, as clarified. We know there are some issues on that compiler but they seem to primarily affect audio. Simply use it if you would like to contribute to testing.

woj1993 commented:
> I found a solution to the strange performance drop. It's just the new changes confusing driver heurestics. The fix is quite extensive, I might pre-merge it as a separate PR.

I have a question. Wouldn't it be easier to contact Nvidia, AMD and Inter. Reporting to them bug in haurestic?

AniLeo commented:
> > I found a solution to the strange performance drop. It's just the new changes confusing driver heurestics. The fix is quite extensive, I might pre-merge it as a separate PR.
> 
> I have a question. Wouldn't it be easier to contact Nvidia, AMD and Inter. Reporting to them bug in haurestic?

We have no way to contact NVIDIA (or Intel), and even then it would probably take too long to reach a solution.

kd-11 commented:
> I have a question. Wouldn't it be easier to contact Nvidia, AMD and Inter. Reporting to them bug in heurestic?

A. It would take months for a resolution.
B. It won't scale going forward, a better solution is always a unified one.

kd-11 commented:
> Can you explain it to somebody that don't have deep knowledge of rendering pipeline?

NV driver sees a big chunk of VRAM that I am using on both CPU and GPU which has flags set to mean it can be either used for computing stuff on GPU then reading results on CPU or uploading stuff from CPU and then using the input for computer (compute then download vs upload then compute). The driver is tuned to guess the first scenario which means the memory on that slab is kept in system RAM, but I need it for the second scenario which means my writes must be immediately flushed to the GPU. That's typically what game-ready drivers do, change such internal guesswork calibration.
In this case we have a better option since most modern systems ship with ReBAR/SAM. Just tell the driver that we're writing straight into VRAM from CPU and the issue goes away.

woj1993 commented:
> > Can you explain it to somebody that don't have deep knowledge of rendering pipeline?
> 
> NV driver sees a big chunk of VRAM that I am using on both CPU and GPU which has flags set to mean it can be either used for computing stuff on GPU then reading results on CPU or uploading stuff from CPU and then using the input for computer (compute then download vs upload then compute). The driver is tuned to guess the first scenario which means the memory on that slab is kept in system RAM, but I need it for the second scenario which means my writes must be immediately flushed to the GPU. That's typically what game-ready drivers do - they just change such internal guesswork calibration. In this case we have a better option since most modern systems ship with ReBAR/SAM. Just tell the driver that we're writing straight into VRAM from CPU and the issue goes away.

Thanks for explanation. Also I tested Asura's Wrath and issues are still present on MSVC buid. So this is all I can tell. My system specs are:

kd-11 commented:
Sadly other drivers are not happy about me abusing ReBAR for common data. I need to narrow down which pools exactly require the setting.
So far, confirmed poor performance on:
- NV driver
- AMDVLK
- RADV

Seems to happen when we have a lot of draw calls, so I need to optimize for the KZ3 edge case. I'm not really concerned about visual glitches, those will get fixed before merge.

digant73 commented:
> @digant73 I fixed the massive latency spikes plaguing nvidia cards, please retest. From my testing, it is now 10% ahead of master in my test scene for KZ3.

it seems better than before although still slightly worse on avg (65 FPS on PR, 67 FPS on master). It seems peak was a little bit better on PR (73 against 70). but I didn't deep testing

kd-11 commented:
> ok, I will make some testing (probably tomorrow) changing those param values on each pool, one by one, and see which one provides an improvement, if any

I already checked and I know some have big difference on NV. Especially vertex-layout, fragment-constants and fragment-textures pools have reasonable bump.
The difference on linux is even greater: 40fps vs 55fps.
High fps tests (400+ fps) rebar is slower than default, but with normal fps < 120 it seems rebar is coming out on top. Definitely needs to be a toggle.

kd-11 commented:
I've added a toggle to UI and removed rebar flag on some heaps where it doesn't help performance in KZ3 but hurts fps in other games like R&C ToD. We can fine tune which heap gets what later, as long as performance is decent for now (matching at least previous master, or within 5% for the worst case)

digant73 commented:
> I've added a toggle to UI and removed rebar flag on some heaps where it doesn't help performance in KZ3 but hurts fps in other games like R&C ToD. We can fine tune which heap gets what later, as long as performance is decent for now (matching at least previous master, or within 5% for the worst case)

I will test tomorrow night

willowdale16 commented:
> > Rebar stuff doesn't affect iGPU right So enable or disable the settings isn't gonna lower/increase the perf?
> 
> It really depends on what the driver is doing with page configuration, but if configured properly, igp should be unaffected. In fact, only NV driver seems to be so badly affected here. Other driver vendors perform the same or slightly better with rebar off.

I can confirm it doesn't affect latest AMD driver on windows (at least for my setup).

digant73 commented:
tested also R FOM. PR seems better (also with my changes). I also got some device-lost after some time (I simply wait on a scene).
The crash is not present on current release.

[RPCS3.log](https://github.com/user-attachments/files/21977569/RPCS3.log)

<img width="3199" height="564" alt="image" src="https://github.com/user-attachments/assets/9c6447ee-12dd-45a8-a7f1-891d66f7c7da" />

digant73 commented:
> > made some custom testing on KZ3. In my case it seems better performance (on par with current release or slightly better) are obtained using `heap_pool_low_latency` also for the following pools:
> > 
> > * `m_attrib_ring_info`
> > * `m_transform_constants_ring_info`
> 
> I found forcing low-latency on vertex inputs to be worse in other games such as ratchet and clank ToD and possibly other games in the future series. Try out some more games, though if the difference is minor anyway, we can just proceed as-is.

which is the name of that pool? I see only:
- `m_vertex_env_ring_info`
- `m_vertex_layout_ring_info`

digant73 commented:
@kd-11 you merged the latest branch master into your PR. But consider that PR #17437 (now merged in your PR) is also hugely regressing performance on some games (FOM for example) as reported in PR #17455 that was basically restoring the performance. Current branch will make difficult again to compare improvements or regressions. For me, It was more logical and safe to merge #17455 (a simple hotfix), and waiting for #17457 (a rework) is mature (and fully tested by the community) to replace the current logic

kd-11 commented:
> @kd-11 you merged the latest branch master into your PR. But consider that PR #17437 (now merged in your PR) is also hugely regressing performance on some games (FOM for example) as reported in PR #17455 that was basically restoring the performance. Current branch will make difficult again to compare improvements or regressions. For me, It was more logical and safe to merge #17455 (a simple hotfix), and waiting for #17457 (a rework) is mature (and fully tested by the community) to replace the current logic

I've rolled back to older branch as base. However, as far as I'm concerned at this point, this PR is past performance testing stage, I'm only focused on the visual problems remaining. It's been open a long time and blocks me from working on anything else.
As for CELL regression, that will get fixed in due time, we don't have to go backwards every time something goes wrong. Remember, there's nothing more permanent in programming than a temporary fix.

kd-11 commented:
> @kd-11 Errors are still happening. For example this: ![Graphical gliches MSVC build commit with fixes](https://private-user-images.githubusercontent.com/11420970/483219365-a9f3e78c-0de3-47fd-80ec-bf2210296835.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0MDcyNDEsIm5iZiI6MTc1NjQwNjk0MSwicGF0aCI6Ii8xMTQyMDk3MC80ODMyMTkzNjUtYTlmM2U3OGMtMGRlMy00N2ZkLTgwZWMtYmYyMjEwMjk2ODM1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODI4VDE4NDkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiYzhjYzFlNWEyYTNlZTI0N2MxOGIwZjhkMGZjZmJjMjQwYWQzMWE4YzgzMmM0YzZmOTE4OGI1NmVlOWUyYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NH4MfRTseQim79tYrtKh5IGhUCFnznMMAopwCL9LHRY) [ ![Graphical gliches MSVC build commit with fixes](https://private-user-images.githubusercontent.com/11420970/483219365-a9f3e78c-0de3-47fd-80ec-bf2210296835.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0MDcyNDEsIm5iZiI6MTc1NjQwNjk0MSwicGF0aCI6Ii8xMTQyMDk3MC80ODMyMTkzNjUtYTlmM2U3OGMtMGRlMy00N2ZkLTgwZWMtYmYyMjEwMjk2ODM1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODI4VDE4NDkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiYzhjYzFlNWEyYTNlZTI0N2MxOGIwZjhkMGZjZmJjMjQwYWQzMWE4YzgzMmM0YzZmOTE4OGI1NmVlOWUyYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NH4MfRTseQim79tYrtKh5IGhUCFnznMMAopwCL9LHRY) ](https://private-user-images.githubusercontent.com/11420970/483219365-a9f3e78c-0de3-47fd-80ec-bf2210296835.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0MDcyNDEsIm5iZiI6MTc1NjQwNjk0MSwicGF0aCI6Ii8xMTQyMDk3MC80ODMyMTkzNjUtYTlmM2U3OGMtMGRlMy00N2ZkLTgwZWMtYmYyMjEwMjk2ODM1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODI4VDE4NDkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiYzhjYzFlNWEyYTNlZTI0N2MxOGIwZjhkMGZjZmJjMjQwYWQzMWE4YzgzMmM0YzZmOTE4OGI1NmVlOWUyYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NH4MfRTseQim79tYrtKh5IGhUCFnznMMAopwCL9LHRY) [ ](https://private-user-images.githubusercontent.com/11420970/483219365-a9f3e78c-0de3-47fd-80ec-bf2210296835.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0MDcyNDEsIm5iZiI6MTc1NjQwNjk0MSwicGF0aCI6Ii8xMTQyMDk3MC80ODMyMTkzNjUtYTlmM2U3OGMtMGRlMy00N2ZkLTgwZWMtYmYyMjEwMjk2ODM1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODI4VDE4NDkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiYzhjYzFlNWEyYTNlZTI0N2MxOGIwZjhkMGZjZmJjMjQwYWQzMWE4YzgzMmM0YzZmOTE4OGI1NmVlOWUyYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NH4MfRTseQim79tYrtKh5IGhUCFnznMMAopwCL9LHRY) Or I once had colorful ship (like rainbow colors not normal textured ship) but I was unable to capture it as it happened after recording was done. [RPCS3.log.gz](https://github.com/user-attachments/files/22028101/RPCS3.log.gz) For you information I won't have access to computer from 31.08 to 14.09. So I won't be able to check this in these days.

That's ok, I can still merge and fix the bugs later if more users report more broken games. Having a sample count of 1 is not good since I can't do anything if I can't see the issue on my end.

digant73 commented:
the device-lost is still present in FOM

[RPCS3.log](https://github.com/user-attachments/files/22031862/RPCS3.log)

digant73 commented:
it's random during the gameplay. Even doing nothing., e.g. starting chapter 3 (`Manchester`) as in the attached video

https://github.com/user-attachments/assets/cafbf340-878d-4fad-b269-100f97fc64ee

kd-11 commented:
> it's random during the gameplay. Even doing nothing, e.g. at the beginning of chapter 3 (`Manchester`) as in the attached video
> 
Try running with validation enabled, hopefully we catch something. I just played two chapters and nothing crashed. No new validation errors either.
The only difference is that my card is older and slower (3050) and I'm on driver 580.97

digant73 commented:
still device-lost after some time

[RPCS3.log](https://github.com/user-attachments/files/22051643/RPCS3.log)

<img width="2932" height="551" alt="image" src="https://github.com/user-attachments/assets/2b35aae7-3d78-472f-8d16-5db0effe9e85" />

digant73 commented:
> > still device-lost after some time
> > attached also a CDL dump
> > [cdl_dump.yaml](https://github.com/user-attachments/files/22051896/cdl_dump.yaml)
> > [RPCS3.log](https://github.com/user-attachments/files/22051643/RPCS3.log)
> 
> You did not enable debug output for validation. CDL is nice but just confirms that the descriptors have a problem with one of the parameters. Validation may give extra info on what that is at the cost of some performance. With the recent binding model the performance hit from validation layers shouldn't be too crazy on such a game.

yes, I used cdl. I will try with validation (I just need to remember how I used it)

digant73 commented:
still crashing with a device-lost (got 3 of them). No crash with the asserts you provided.
Attached the log
[RPCS3_device-lost.log](https://github.com/user-attachments/files/22082890/RPCS3_device-lost.log)

digant73 commented:
it seems commit `ad094a0b` (`vk: Move texture parameter stream to SSBO layout`) is responsible for the device-lost. Starting from that commit I was able to get the device-lost.

**NOTE:**
Sine `vk: Move FS context vars to block format` I needed to apply `vk: Fix wpos computation`. Otherwise FOM was crashing at boot due to shader error

digant73 commented on rpcs3/Emu/RSX/VK/VKCommonDecompiler.cpp:
```diff
@@ -20,7 +20,8 @@ namespace vk
 		{ "spec_color1", 13 },
 		{ "fog_c", 14 },
-		{ "fogc", 14 }
+		{ "fogc", 14 },
```
is it correct 14? it is already used for `fog_c`

kd-11 commented:
@digant73 Should be fixed now, retest.

digant73 commented:
> @digant73 Should be fixed now, retest.

it is still present

kd-11 commented:
> > @digant73 Should be fixed now, retest.
> 
> it is still present

Try setting all the heaps to 1GB starting size. I'm like 99% sure the grow mechanism is causing stale references in the descriptors.

digant73 commented:
still crashing even using multiplier 1024 and 2048. Below I used 1024:

```
	// VRAM allocation
	// This first set is bound persistently, so grow notifications are enabled.
	m_attrib_ring_info.create(VK_BUFFER_USAGE_UNIFORM_TEXEL_BUFFER_BIT, VK_ATTRIB_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_default, "attrib buffer", 0x400000, VK_TRUE);
	m_fragment_env_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, VK_UBO_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_low_latency, "fragment env buffer", 1024 * 0x10000, VK_TRUE);
	m_vertex_env_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, VK_UBO_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_low_latency, "vertex env buffer", 1024 * 0x10000, VK_TRUE);
	m_fragment_texture_params_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, VK_UBO_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_low_latency, "fragment texture params buffer", 1024 * 0x10000, VK_TRUE);
	m_vertex_layout_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, VK_UBO_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_low_latency, "vertex layout buffer", 1024 * 0x10000, VK_TRUE);
	m_fragment_constants_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, VK_UBO_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_low_latency, "fragment constants buffer", 1024 * 0x10000, VK_TRUE);
	m_transform_constants_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, VK_TRANSFORM_CONSTANTS_BUFFER_SIZE_M * 0x100000, vk::heap_pool_default, "transform constants buffer", 1024 * 0x10000, VK_TRUE);
	m_raster_env_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, VK_UBO_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_low_latency, "raster env buffer", 1024 * 0x10000, VK_TRUE);
	// Below here, we do not bind these persistently. Each draw call specifies the range manually so we do not need heap_grow notifications.
	m_instancing_buffer_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, VK_TRANSFORM_CONSTANTS_BUFFER_SIZE_M * 0x100000, vk::heap_pool_default, "instancing data buffer");
	m_index_buffer_ring_info.create(VK_BUFFER_USAGE_INDEX_BUFFER_BIT, VK_INDEX_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_default, "index buffer");
	m_texture_upload_buffer_ring_info.create(VK_BUFFER_USAGE_TRANSFER_SRC_BIT, VK_TEXTURE_UPLOAD_RING_BUFFER_SIZE_M * 0x100000, vk::heap_pool_default, "texture upload buffer", 32 * 0x100000);

```

digant73 commented:
it crashes even with this (1024 * 100000):

```
	m_attrib_ring_info.create(VK_BUFFER_USAGE_UNIFORM_TEXEL_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_default, "attrib buffer", 0x400000, VK_TRUE);
	m_fragment_env_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_low_latency, "fragment env buffer", 0x10000, VK_TRUE);
	m_vertex_env_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_low_latency, "vertex env buffer", 0x10000, VK_TRUE);
	m_fragment_texture_params_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_low_latency, "fragment texture params buffer", 0x10000, VK_TRUE);
	m_vertex_layout_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_low_latency, "vertex layout buffer", 0x10000, VK_TRUE);
	m_fragment_constants_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_low_latency, "fragment constants buffer", 0x10000, VK_TRUE);
	m_transform_constants_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_default, "transform constants buffer", 0x10000, VK_TRUE);
	m_raster_env_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_low_latency, "raster env buffer", 0x10000, VK_TRUE);
	// Below here, we do not bind these persistently. Each draw call specifies the range manually so we do not need heap_grow notifications.
	m_instancing_buffer_ring_info.create(VK_BUFFER_USAGE_STORAGE_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_default, "instancing data buffer");
	m_index_buffer_ring_info.create(VK_BUFFER_USAGE_INDEX_BUFFER_BIT, 1024 * 0x100000, vk::heap_pool_default, "index buffer");
	m_texture_upload_buffer_ring_info.create(VK_BUFFER_USAGE_TRANSFER_SRC_BIT, 1024 * 0x100000, vk::heap_pool_default, "texture upload buffer", 32 * 0x100000);

```

digant73 commented:
I also updated drivers to latest version. The issue is present also on other games such as R3 (running at 60 fps). It seems similar to what has been fixed in the past (issue on descriptors). No more device-lost since that bugfix.
I will apply 1024 * 100000 to the commit where I started to see the device-lost. Just to see if it fixes something there

kd-11 commented:
@digant73 Use this branch: https://github.com/kd-11/rpcs3/tree/descriptors-debug
For best results, also have CDL enabled at the same time and attach both logs after the crash. Don't close rpcs3 immediately after crash, let the diagnostics dump complete first (a message will appear in the log when its done).

digant73 commented:
ok. I will test tomorrow

digant73 commented:
attached the requested files

[shaderlog.zip](https://github.com/user-attachments/files/22244614/shaderlog.zip)
[2025-09-10-011914.zip](https://github.com/user-attachments/files/22244615/2025-09-10-011914.zip)

digant73 commented:
here the logs enabling in CDL:
- Synchronize commands
- Instrument all commands
FPS are a lot lower (min. 20 FPS) but the game still easily crashes.

[2025-09-10-173423_FOM_10.zip](https://github.com/user-attachments/files/22258196/2025-09-10-173423_FOM_10.zip)

<img width="1270" height="1070" alt="image" src="https://github.com/user-attachments/assets/5c6dd36d-6a38-4823-a2d7-6298164b63e2" />

[shaderlog.zip](https://github.com/user-attachments/files/22258194/shaderlog.zip)

kd-11 commented:
@digant73 Thanks.
The CDL dump does point out a particular image as being the possible culprit.

```yaml
type: Invalid Read
        begin: 0x00000003EACAC000
        end: 0x00000003EACACFFF
        matchingAddressRecords:
          - begin: 0x00000003EA674000
            end: 0x00000003EAE94000
            type: VkImage
            handle: 0x000001DC9DD1B470[]
            currentlyBound: false
 ```          
 With synchronized draws we see the crash is indeed from a specific shader draw command. Unfortunately I failed to account for shader cache when building pipeline data so I'll have to share a separate patch (or you can re-run with shader cache cleared)

kd-11 commented:
Thanks. I see what is happening. I believe we have found the cause - that image read is borked.
I unfortunately don't have enough info from CDL to tell me exactly which shader/descriptor-set crashed, but I can say this crash has nothing to do with descriptor bind itself. Its an out-of-bounds access crash.
We have robustness enabled, so buffer access shouldn't crash, but I suspect invalid pointer math could lead to a crash, need to think about this a bit.

digant73 commented:
so no CDL, no cleared cache, no Log Shader programs and Renderdoc Compatibility Mode this time?
Only `rpcs3.log` and `rpcs3.pdb` this time?
Ues I simply compile your code and the execute it from the `bin` output folder

digant73 commented:
here the crash with CDL
Rename the `RPCS3.zip` to `RPCS3.7z`

[RPCS3.zip](https://github.com/user-attachments/files/22312207/RPCS3.zip)

[2025-09-13-194222_FOM_20.zip](https://github.com/user-attachments/files/22312209/2025-09-13-194222_FOM_20.zip)

kd-11 commented:
This time the handle is 100% valid during the crash. So we've fixed some bugs and circled right back to the start. There's no path forward at this time. I'll PR the invalidation bugfix, it probably fixes some other crashes on master.

woj1993 commented:
@kd-11 Can you rebase your branch on 3e701a7 or merge newest main branch commits. Issues with sound were fixed on master in latest commit. Apparently on windows we had broken "stack frame"

digant73 commented:
> @digant73 Can you download latest commit and retest after removing cache folder from rpcs3 directory? This way we will confirm if it is still happening after stack frame corruption has been fixed. I will check Asura's Wrath in meantime.

it is crashing even on latest commit

Jonathan44062 commented:
Got VK_ERROR_DEVICE_LOST
Resistance: Fall of Man
[RPCS3.log](https://github.com/user-attachments/files/22585695/RPCS3.log)

Jonathan44062 commented:
Tested the new commit, same crash
[RPCS3.log](https://github.com/user-attachments/files/22586520/RPCS3.log)

kd-11 commented:
> @kd-11 as previously reported (27 days ago) I started to get the device-lost since commit [ad094a0](https://github.com/RPCS3/rpcs3/commit/ad094a0b9d7f08d9c01cce6e61d19aef0d124a1a) (vk: Move texture parameter stream to SSBO layout). Is maybe possible to temporary rollback (just for testing) only this feature in the PR and test if the issue is still present or not? Just to narrow down the cause

Rolling back that one commit changes nothing. It doesn't tell us why the driver is crashing for some users despite no spec violations. There is nothing wrong with the commit itself, the texture parameter buffer just has scaling coefficients for the sampler operations. Even if we fed it garbage or nothing at all it still shouldn't crash due to memory access guarantees + robustness feature.

digant73 commented:
> > @kd-11 as previously reported (27 days ago) I started to get the device-lost since commit [ad094a0](https://github.com/RPCS3/rpcs3/commit/ad094a0b9d7f08d9c01cce6e61d19aef0d124a1a) (vk: Move texture parameter stream to SSBO layout). Is maybe possible to temporary rollback (just for testing) only this feature in the PR and test if the issue is still present or not? Just to narrow down the cause
> 
> Rolling back that one commit changes nothing. It doesn't tell us why the driver is crashing for some users despite no spec violations. There is nothing wrong with the commit itself, the texture parameter buffer just has scaling coefficients for the sampler operations. Even if we fed it garbage or nothing at all it still shouldn't crash due to memory access guarantees + robustness feature.

Yes, it is a change just for testing if the crash will be no more present with all the other features on this PR still present.

digant73 commented:
I tried to use `nsight grapjics` but it seems it doesn't allow to save/export when rpcs3 crashes.
However, I was able to previously use `nsight systems` and it allowed to save a big log (30GB unzipped, 3GB zipped).
Below the link for download, if it can be useful.

https://mega.nz/file/fipEEIqJ#KMJjh6Ya_9k4GmDYvCFmUBhm-wr4jGTFL1Y63HBC3wA

kd-11 commented:
@digant73 We need two things:
1. Enable renderdoc compat setting to introduce debug names to the trace.
2. Zip up the shader debug information. According to aftermath notes, it should be a separate folder.
3. Disable MSAA during the test. I want to eliminate some edge cases that are unique to NVIDIA cards.

Findings so far:
- Crash is coming from FS.
- The memory address crashing is valid. It is not a resource not-available or destroyed issue.
- The memory could not be accessed, so we're either reading a write-only slab or vice-versa. Could also be incorrect access type but this is very unlikely.

digant73 commented:
here the dump with Renderdoc compat enabled and MSAA disabled.
The zip contains also the nvidia shader debug information that could be used to produce binary shader (SPIR-V?) and separate shader debug info (.pdb?).

[NVIDIA Corporation.zip](https://github.com/user-attachments/files/22648806/NVIDIA.Corporation.zip)

kd-11 commented:
> Are the crashing problems only isolated to nvidia cards?

@willowdale16 Yes
